File: Python/from_ecote_book/p217_actual_practice2.py
# x 입력
x = int(input())

# 계산된 결과를 저장하기 위한 dp 테이블
d = [0] * 30001

# dp 보텀업 진행
for i in range(2, x+1):
    d[i] = d[i-1] + 1

    if i % 2 == 0:
        d[i] = min(d[i], d[i//2] + 1)
    if i % 3 == 0:
        d[i] = min(d[i], d[i//3] + 1)
    if i % 5 == 0:
        d[i] = min(d[i], d[i//5] + 1)

print(d[x])

File: JavaScript/from_ecote_book/actual_practices/p217_actual_practice2.js
// x 입력
const input = require("fs").readFileSync("/dev/stdin").toString().trim();
const x = parseInt(input);

// 계산된 결과를 저장하기 위한 dp 테이블
const d = new Array(30001).fill(0);

// dp 보텀업 진행
for (let i = 2; i < x + 1; i++) {
  // 1 빼는 경우
  d[i] = d[i - 1] + 1;

  // 2로 나누는 경우
  if (i % 2 === 0) {
    d[i] = Math.min(d[i], d[Math.floor(i / 2)] + 1);
  }
  // 3으로 나누는 경우
  if (i % 3 === 0) {
    d[i] = Math.min(d[i], d[Math.floor(i / 3)] + 1);
  }
  // 5로 나누는 경우
  if (i % 5 === 0) {
    d[i] = Math.min(d[i], d[Math.floor(i / 5)] + 1);
  }
}

console.log(d[x]);
